netutils/forward.go

```diff
@@ -23,102 +22,66 @@ type Conn interface {
-	// process.
+	err = netns.Set(dst)
+	if err != nil {
+		return errors.Wrapf(err, "fail to set current namespace to dst %v", dst)
+	}
 
-	// So we need to pass through an intermediary IPC system, here a UNIX socket.
-	// Basically the parent process is creating a unix socket server, the child
-	// process in the NS of our target is creating a connection to this socket,
-	// this connection is linking the remote client connection and the internal
-	// sand connection.
-	//
-	// HTTP client <----- inter-process unix conn -----> connection to target in sand network
-	unixSocket, err := net.ListenUnix("unix", &net.UnixAddr{Net: "unix", Name: addr})
+	dstHost := fmt.Sprintf("%s:%s", ip, port)
+	dstSocket, err := net.Dial("tcp", dstHost)
 	if err != nil {
-		return errors.Wrapf(err, "fail to open unix socket")
+		return errors.Wrapf(err, "fail to open connection to %v", dstHost)
 	}
-	defer unixSocket.Close()
 
+	wg := &sync.WaitGroup{}
+	wg.Add(2)
+
+	// Error logs are Info as we're only proxying connections, if one is stopped abruptely
+	// The real error is either from the client side or the destination side, it's not an
+	// error for the proxy itself
 	go func() {
-		clientSocket, err := unixSocket.AcceptUnix()
-		if err != nil {
-			log.WithError(err).Error("fail to accept unix connection")
+		defer wg.Done()
+		defer dstSocket.Close()
+		_, err := io.Copy(dstSocket, srcSocket)
+		if err != io.EOF {
+			log.Info("end of connection from unix src to dst socket with error: %v", err)
 			return
 		}
-		log.Info("socket unix connection accepted from child process")
-		defer clientSocket.Close()
-
-		wg := &sync.WaitGroup{}
-		wg.Add(2)
-		go func() {
-			defer wg.Done()
-			defer clientSocket.CloseWrite()
-			defer src.Close()
-			_, err := io.Copy(clientSocket, src)
-			if err != nil && err != io.EOF {
-				log.WithError(err).Error("fail to copy data from HTTP src connection to unix socket")
-				return
-			}
-			log.Info("end of connection from HTTP src connection to unix socket")
-		}()
-
-		go func() {
-			defer wg.Done()
-			defer src.CloseWrite()
-			defer clientSocket.Close()
-			_, err := io.Copy(src, clientSocket)
-			if err != nil && err != io.EOF {
-				log.WithError(err).Error("fail to copy data from unix socket to HTTP src connection")
-				return
-			}
-			log.Info("end of connection from unix socket to HTTP src connection")
-		}()
-
-		wg.Wait()
+		log.Debug("end of connection from unix src to dst socket")
 	}()
 
-	cmd := &exec.Cmd{
-		Path:   reexec.Self(),
-		Args:   append([]string{"sc-netns-pipe-socket"}, ns, ip, port, addr),
-		Stderr: os.Stderr,
-		Stdout: os.Stdout,
-	}
+	go func() {
+		defer wg.Done()
+		defer srcSocket.Close()
+		_, err := io.Copy(srcSocket, dstSocket)
+		if err != io.EOF {
+			log.Info("end of connection from dst socket to src socket with error: %v", err)
+			return
+		}
+		log.Debug("end of connection from dst socket to src socket")
+	}()
+	wg.Wait()
 
-	err = cmd.Run()
+	err = netns.Set(current)
```

Comment:
Shouldn't we do this in a defer, right after the line `err = netns.Set(dst)`?

Reply:
Actually I've a doubt the defer func()  would respect the runtime.LockOsThread(), as it seams, it's really local to the current goroutine, I'm not sure the `defer func()` is part of it. What do you think?

Reply:
That's a good question. From my understanding, the current goroutine is the one executing `ForwardConnection`. And that's the one which executes the `netns.Set(current)` function. When calling `defer`, you are still executing in the same goroutine. Hence it should not be an issue with the call to `runtime.LockOsThread`.

Reply:
Tested in a minimal go program, defer seems to work as expected

```
package main

import (
	"fmt"
	"net"
	"runtime"

	"github.com/vishvananda/netns"
)

func main() {
	// Lock the OS Thread so we don't accidentally switch namespaces
	runtime.LockOSThread()
	defer runtime.UnlockOSThread()

	// Save the current network namespace
	origns, err := netns.Get()
	fmt.Println(origns.UniqueId(), err)
	defer origns.Close()

	// Create a new network namespace
	newns, err := netns.New()
	fmt.Println(newns.UniqueId(), err)
	defer newns.Close()

	// Switch back to the original namespace
	defer func() {
		ifaces, _ := net.Interfaces()
		fmt.Printf("Interfaces: %v\n", ifaces)

		err = netns.Set(origns)
		fmt.Println(err)

		ifaces, _ = net.Interfaces()
		fmt.Printf("Interfaces: %v\n", ifaces)
	}()
}
```